File: ringbuffer-go/ringbuffer.go
package main

import (
  "fmt"
  "errors"
)

type RingBuffer struct {
  size uint
  readCursor uint
  writeCursor uint
  data []int
}

func createBuffer(size int) *RingBuffer {
  r := &RingBuffer{}
  r.data = make([]int, size, size)
  return r
}

func (buf *RingBuffer) append(item int) bool {
  if buf.size < uint(len(buf.data)) {
    buf.data[buf.writeCursor] = item
    buf.size +=1
    buf.writeCursor = (buf.writeCursor+1) % uint(len(buf.data))
    return true
  }
  return false
}

func (buf *RingBuffer) Write(items []int) int {
  items_written := 0
  for _, v := range items {
    if buf.append(v) {
      items_written += 1
    } else {
      break
    }
  }
  return items_written
}

func (buf RingBuffer) IsEmpty() bool {
  return buf.size == 0
}

func (buf *RingBuffer) Read() (int, error) {
  if buf.IsEmpty() {
    return 0, errors.New("ringbuffer is empty")
  }
  value := buf.data[buf.readCursor]
  buf.readCursor = (buf.readCursor+1) % uint(len(buf.data))
  buf.size -= 1
  return value, nil
}

func main() {
  buf := createBuffer(5)
  i := buf.Write([]int{1,2,3})
  fmt.Printf("%d items were written\n", i)
  i = buf.Write([]int{4,5,6})
  fmt.Printf("%d items were written\n", i)
  var err error
  var v int
  for !buf.IsEmpty() {
    v, err = buf.Read()
    if err == nil {
      fmt.Printf("read %d\n",v)
    }
  }
  fmt.Println("Finish reading")
}

File: ringbuffer.py
import sys

class RingBuffer:
    def __init__(self, maxlen):
        self.q = [None]*maxlen
        self.size = 0
        self.writeCursor = 0
        self.readCursor = 0

    def append(self, item):
        if self.size < len(self.q):
            self.q[self.writeCursor] = item
            self.size += 1
            self.writeCursor = (self.writeCursor + 1) % len(self.q)
            return True
        return False

    def write(self, items):
        total_added = 0
        for i in items:
            if self.append(i):
                total_added += 1
            else:
                break
        return total_added

    def read(self):
        if self.size > 0:
            self.size -= 1
            val = self.q[self.readCursor]
            self.readCursor = (self.readCursor+1) % len(self.q)
            return val

        raise BaseException("ringbuffer is empty")

    def __str__(self):
        return "{} r={},w={},size={}".format(self.q, self.readCursor, self.writeCursor, self.size)

if __name__ == '__main__':

    try:
        buffer = RingBuffer(5)

        print("Writing [1,2,3] to ringbuffer")
        items_written = buffer.write([1, 2, 3])
        print("{} items were written".format(items_written))

        print("Writing [4,5,6] to ringbuffer")
        items_written = buffer.write([4, 5, 6])
        print("{} items were written".format(items_written))

        print("Writing [100,100,100] to ringbuffer")
        items_written = buffer.write([100, 100, 100])
        print("{} items were written".format(items_written))

        print("reading ", buffer.read())
        print("reading ", buffer.read())

        print("Writing [7,77,777,7777] to ringbuffer")
        items_written = buffer.write([7, 77, 777, 7777])
        print("{} items were written".format(items_written))


        print("reading ", buffer.read())
        print("reading ", buffer.read())
        print("reading ", buffer.read())
        print("reading ", buffer.read())
        print("reading ", buffer.read())
    except:
        print(str(sys.exc_info()[0]))
